pkg/job/job.go

```diff
@@ -0,0 +1,218 @@
+			logrus.Errorln(err)
+		}
+
+		j.changeState(proto.ProcessState_EXITED)
+
+		// Inform stop caller to finish.
+		// The channel buffer size is 1, so if this job exit naturally, it won't be blocked
+		j.exitChan <- true
+		close(j.exitChan)
+	}()
+
+	return nil
+}
+
+// Stop terminates a job. If the force flag is set to true, it will using SIGKILL to terminate the process, otherwise it will be SIGTERM
+func (j *Impl) Stop(force bool) error {
+	if j.getState() != proto.ProcessState_RUNNING || j.cmd.Process == nil {
+		return ErrNotRunning
+	}
+
+	// Use SIGKILL to force the process to stop immediately, otherwise we will use SIGTERM
+	if force {
+		if err := j.cmd.Process.Signal(os.Kill); err != nil {
+			return err
+		}
+	} else {
+		if err := j.cmd.Process.Signal(syscall.SIGTERM); err != nil {
+			return err
+		}
+	}
+
+	// Wait for the job to exit with timeout of 1s
+	select {
+	case _, ok := <-j.exitChan:
+		if !ok {
+			// exit channel is already closed, which means the process is already exited
+			return ErrNotRunning
+		}
+	case <-time.After(time.Second):
+		return ErrTimeout
+	}
+
+	j.changeState(proto.ProcessState_STOPPED)
+
+	return nil
+}
+
+// Status queries the current statis of the job
+func (j *Impl) Status() *proto.JobStatus {
+	state := j.getState()
+	pid := -1
+	exitcode := -1
+
+	if j.cmd.Process != nil {
+		pid = j.cmd.Process.Pid
+	}
+
+	// if state > proto.ProcessState_RUNNING, the process is already terminated and ProcessState can be read safely
+	if state > proto.ProcessState_RUNNING && j.cmd.ProcessState != nil {
+		exitcode = j.cmd.ProcessState.ExitCode()
+	}
+
+	return &proto.JobStatus{
+		Job:     &proto.Job{Id: j.id.String()},
+		Command: &proto.Command{Cmd: j.String()},
+		Owner:   j.owner,
+		Status: &proto.ProcessStatus{
+			Pid:      int32(pid),
+			State:    state,
+			ExitCode: int32(exitcode),
+		},
+	}
+}
+
+// String returns the command wrapped by the job
+func (j *Impl) String() string {
+	return j.cmd.String()
+}
+
+// getState reads current job state with mutex read protection
+func (j *Impl) getState() proto.ProcessState {
+	j.stateMutex.RLock()
+	defer j.stateMutex.RUnlock()
+
+	return j.state
+}
+
+// changeState sets the new state of the job
+func (j *Impl) changeState(state proto.ProcessState) {
+	j.stateMutex.Lock()
+	defer j.stateMutex.Unlock()
+
+	j.state = state
+}
+
+// configNameSpace applies namespaces policy on the process
+func configNameSpace() *syscall.SysProcAttr {
+	// TODO Continue to reinforce namespaces
+	return &syscall.SysProcAttr{
+		Cloneflags: syscall.CLONE_NEWNS |
+			syscall.CLONE_NEWUTS |
+			syscall.CLONE_NEWIPC |
+			syscall.CLONE_NEWPID |
+			syscall.CLONE_NEWNET |
+			syscall.CLONE_NEWUSER,
+		UidMappings: []syscall.SysProcIDMap{
+			{
+				ContainerID: 0,
+				HostID:      os.Getuid(),
+				Size:        1,
+			},
+		},
+		GidMappings: []syscall.SysProcIDMap{
+			{
+				ContainerID: 0,
+				HostID:      os.Getgid(),
+				Size:        1,
+			},
+		},
+	}
+}
```

Comment:
Yeah I know, but can I keep it?

Reply:
Part of the challenge is to keep your implementation simple. We have told you this is not necessary, I don't think you have explained why you would like to keep it

Reply:
There is a test case where sudo is passed in command, without user namespace isolation the process is allow to call sudo, it won't do anything dangerous if we don't give the right password, but it can launch the tty and halt the program. If you think I should remove it then I will remove it and the sudo test case